File: app/src/main/java/de/metzgore/beansplan/dailyschedule/DailyScheduleViewModel.java
package de.metzgore.beansplan.dailyschedule;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.NonNull;

import java.util.Date;

import de.metzgore.beansplan.data.room.relations.DailyScheduleWithShows;
import de.metzgore.beansplan.data.room.Reminder;
import de.metzgore.beansplan.data.room.Show;

public class DailyScheduleViewModel extends ViewModel {

    private final MutableLiveData<Date> dateToLoad = new MutableLiveData<>();
    private final LiveData<DailyScheduleWithShows> schedule;
    private final DailyScheduleRepository repo;
    public final LiveData<Boolean> isEmpty;

    public DailyScheduleViewModel(DailyScheduleRepository scheduleRepo) {
        repo = scheduleRepo;
        schedule = Transformations.switchMap(dateToLoad, repo::loadScheduleFromCache);
        isEmpty = Transformations.map(schedule, schedule -> schedule == null || schedule.shows
                .isEmpty());
    }

    public LiveData<DailyScheduleWithShows> getSchedule() {
        return schedule;
    }

    public void loadSchedule(@NonNull Date date) {
        dateToLoad.setValue(date);
    }

    public void upsertReminder(Show show, Reminder reminder) {
        repo.upsertReminder(show, reminder);
    }

    public void deleteReminder(Show show, Reminder reminder) {
        repo.deleteReminder(show, reminder);
    }
}

File: app/src/main/java/de/metzgore/beansplan/dailyschedule/DailyScheduleViewModel.kt
package de.metzgore.beansplan.dailyschedule

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import de.metzgore.beansplan.data.room.relations.DailyScheduleWithShows
import de.metzgore.beansplan.data.room.relations.ShowWithReminder
import java.util.*

class DailyScheduleViewModel(private val repo: DailyScheduleRepository) : ViewModel() {

    private val dateToLoad = MutableLiveData<Date>()
    val schedule: LiveData<DailyScheduleWithShows> = Transformations.switchMap(dateToLoad) { repo.loadScheduleFromCache(it) }
    val isEmpty: LiveData<Boolean> = Transformations.map(schedule) { schedule ->
        schedule == null || schedule.shows.isEmpty()
    }

    fun loadSchedule(date: Date) {
        dateToLoad.value = date
    }

    fun upsertReminder(showWithReminder: ShowWithReminder) {
        repo.upsertReminder(showWithReminder)
    }

    fun deleteReminder(showWithReminder: ShowWithReminder) {
        repo.deleteReminder(showWithReminder)
    }
}
